protocols/abci/addressing-1174.md

```diff
@@ -0,0 +1,230 @@
+> and `ProcessProposal` may break Requirements 3 to 5.
+
+## Solution Proposed
+
+### Modified consensus _validity_
+The specification of consensus implicitly considers `valid(v)` as an immutable function,
+which is always supposed to provide the same result, when called at the same height
+for the same value `v`.
+This was the main reason for introducing the determinism requirements on `ProcessProposal`
+in the ABCI 1.0 specification.
+
+If we are to relax the determinism requirements on the application,
+we first need to weaken the `valid(v)` predicate and therefore consensus _validity_.
+In the proposed solution, correct processes MAY evaluate `valid(v)` to _true_ or _false_ at different times
+for the same value `v` during height `h`.
+
+Consensus _validity_ invariant is then modified as follows:
+
+- _weak validity_: predicate `valid(v)` has been evaluated to _true_ at least once by one correct process
+  for the decided block `v`.
+
+### Eventual requirements
+
+We now relax the relevant ABCI 1.0 requirements in the following way.
+
+* Requirement 3b [`PrepareProposal`, `ProcessProposal`, eventual coherence]:
+  There exists a round $r_s \ge 0$ of height $h$ such that,
+  for any two correct processes $p$ and $q$ and any round $r_p \geq r_s$,
+  if $q$'s CometBFT calls `RequestProcessProposal` on $u_p$,
+  $q$'s application returns Accept in `ResponseProcessProposal`.
+
+* The determinism-related requirements, namely requirements 4 and 5, are removed.
+
+We call round $r_s$ the coherence-stabilization round.
+
+> If we think in terms of `valid(v, bc(h-1), x_p)`, observe that it is in the responsibility of the application to
+> ensure 3b, that is, the application designers need to prove that the `x_p` values at correct processes 
+> are over time evolving
+> in a way that eventually ResponseProcessProposal returns ACCEPT.
+> In analogy to proposer based time, `x_p` can be considered to be process `p`s local clock, and having
+> clocks synchronized was used to prove acceptance of a proposal.
+
+> [TODO: can we _tighten up_ eventual coherence?... i.e., make it even weaker?]
+>
+> [TODO: Since we have removed determinism, can byzantine proposers cause mayhem now?]
+
+### Modifications to the consensus Algorithm
+
+The Tendermint algorithm as described in the [arXiv paper][arxiv],
+and as implemented in CometBFT up to version `v0.38.0-rc3`,
+cannot guarantee consensus _termination_ for applications
+that just fulfill requirement 3b (eventual coherence), but do not fulfill requirements 3, 4, and 5 .
+
+We need the following modifications (in terms of the algorithm as described in page 6 of the arXiv paper):
+
+- remove the evaluation of `valid(v)` in lines 29, 36 and 50 (i.e. replace `valid(v)` by `true`)
+- modify line 23 as follows
+
+> _\[Original\]_ &nbsp; 23: **if** $valid(v) \land (lockedRound_p = −1 \lor lockedValue_p = v)$ **then**
+
+> _\[Modified\]_
+>
+> &nbsp; 23a: $validValMatch := (validRound_p \neq -1 \land validValue_p = v)$
```

Comment:
So, if `validRound_p == -1` then necessarily `lockedRound_p == -1`.

Considering all the values these pair of variables (valid and locked) can have during the execution, we have that locked values/rounds are subset of valid values/rounds. Namely, `locked(v, r) => valid(v, r)`.

Reply:
Can we use that info to simplify the expression?

Reply:
So, if $validValue_p =v$ then necessarily $validRound_p \geq 0$. Namely, the first part is redundant.

Like I suggested above, I would use this additional term to define the call to $valid(x)$, namely, either  $validValue_p =v$  or we run $valid(v)$.